components/board/customlinux/setup.go

```diff
@@ -0,0 +1,83 @@
+// Package customlinux implements a board running linux
+package customlinux
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/pkg/errors"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/board/genericlinux"
+)
+
+// GenericLinuxPin describes a gpio pin on a linux board.
+type GenericLinuxPin struct {
+	Name            string `json:"name"`
+	Ngpio           int    `json:"ngpio"`
+	RelativeID      int    `json:"relative_id"`
+	PWMChipSysFSDir string `json:"pwm_chip_dir,omitempty"`
+	PWMID           int    `json:"pwm_id,omitempty"`
+}
+
+// GenericLinuxPins describes a list of pins on a linux board.
+type GenericLinuxPins struct {
+	Pins []GenericLinuxPin `json:"pins"`
+}
+
+// UnmarshalJSON handles setting defaults for pin configs.
+func (conf *GenericLinuxPin) UnmarshalJSON(text []byte) error {
+	type TempPin GenericLinuxPin
+	aux := TempPin{
+		Ngpio:      -1,
+		RelativeID: -1,
+		PWMID:      -1,
+	}
+	if err := json.Unmarshal(text, &aux); err != nil {
+		return err
+	}
+	*conf = GenericLinuxPin(aux)
+	return nil
+}
+
+// Validate ensures all parts of the config are valid.
+func (conf *GenericLinuxPin) Validate(path string) error {
+	if conf.Name == "" {
```

Comment:
If `GenericLinuxPins` contained a map from names to pin definitions that didn't have names, this check would be unnecessary. If convenient, try to design data structures so that impossible states are impossible to represent.